### ISSUE TITLE
Portal into protected area

### ISSUE THREAD

**Issue by Bastrabun** [minetest-mods/nether#26]
Currently it is possible to activate a portal on one side and let the corresponding portal create itself in a protected area.

Can you add a is_protected check to where the new portal is created?

**Comment by Treer** [comment]
Will it be enough to check is_protected() for each corner of the portal's bounding box, rather than on every node inside the bounding box?

The protection mod I've used added protection in bounding boxes, plus doors and chests, and doors and chests should already fail the volume_is_natural() check, but I know there's more than one protection mod and I'm not familiar with them all.

**Comment by Bastrabun** [comment]
It is still possible to create a portal into a protected area. I can show you if you want

**Comment by Treer** [comment]
yes, you'll need to show me. I don't know what protection mod is being used.

It should be possible to create a portal into a protected area if it's your protected area, or if a working portal frame already exists there.

**Comment by Treer** [comment]
Could you set the DEBUG_MODE in nether/init.lua to 1 or 4 and capture the result.
Specificaly I'd be looking when whether it logs 
`Volume is protected ...`
vs
`Volume is natural and unprotected for player ...`

when you open a portal into a protected area

**Comment by Bastrabun** [comment]
Wow, quick response. If you want I can show you live? Join server at address your-land.de port 30000

I am using the areas mod https://github.com/minetest-mods/areas

Until then I will try as you suggested. I set the debug to 4 to get a nice log and I inserted into portal_api.lua this:

`minetest.log("action","called. minp="..dump(minp)..",maxp="..dump(maxp)..",player_name="..dump(player_name))`

I can see a good many calls to the function in the log from my makeshift insertion above, while debug = 4 does not write to the log, although I enabled log level info. So I tried to see it with debug = 1

```
Volume is protected (2082,-272,1162)-(2087,-268,1167)
No destination_orientation given
Destination set to (2084,-272,1164)
set_portal_metadata(ignite=true) at (2084,-26501,1164) orient 90, setting to target (2084,-271,1165)
Adding/updating portal in mod_storage: (2084,-26501,1164) is nether_portal
ensure_remote_portal_then_teleport() saw default:stone at (2084,-271,1165) rather than a wormhole. Calling locate_or_build_portal()
locate_or_build_portal() called at wormholePos(2084,-271,1165) with suggested orient 90, targeted to (2084,-26500,1165)
Placed nether_portal portal schematic at (2082,-272,1164), orientation 90
set_portal_metadata(ignite=false) at (2084,-272,1164) orient 90, setting to target (2084,-26500,1165)
Adding/updating portal in mod_storage: (2084,-272,1164) is nether_portal
Teleporting player from wormholePos(2084,-26500,1165) to wormholePos(2084,-271,1165)
the player already teleported from (2084,-26501,1164), and is now standing in a different portal - (2084,-272,1164)
the player already teleported from (2084,-26501,1164), and is now standing in a different portal - (2084,-272,1164)
the player already teleported from (2084,-26501,1164), and is now standing in a different portal - (2084,-272,1164)
portal checkup at 2 seconds
portal checkup at 6 seconds
```

**Comment by Treer** [comment]
Thank you for getting the log, I think I can see the problem from it.

When looking for a portal location, it starts at the surface and checks for unnatural/protected nodes, and if it finds any then it tries again a bit deeper underground etc. After the loop tries and fails 16 attempts it gives up and puts the portal at (y - 256).

I suspect this is what's happening - that area is unnatural or protected all the way down to -272.

I'm not very familiar with the sort of situation that leads to this. What would be a good game-design solution - how common is it to be protected all the way down that deep?

I could extend how deep it will search. I could even make the search increase in depth logarithmically.

**Comment by Bastrabun** [comment]
Especially around the spawn areas can reach quite deep. I'd prefer the portalling fail if the area is protected - no matter by whom - and tell the player that "Mysterious forces prevented you from opening that portal. Please try another location" or something along those lines

Still, other users/serverowners may think otherwise.

What most would definitely want to avoid is a portal being created in something other than air. Even what the function considers "natural" may have been in fact built. One of my players made himself an elaborate tree house :D

**Comment by Treer** [comment]
It can't guarantee air as there's often no way to know where the surface is (attempting to place the portal on the surface instead of defaulting to underground is new and imperfect feature), and a surface is never clear of decoration, but if a player has made a tree hut they've almost certainly used some nodes that will fail the natural test, so it should be safe (perhaps if spawned at the right distance it could damage a small corner of a tree but repairing that should be a trivial one time effort).

I can make it check deeper, and I should be able to make it fail to ignite rather than have a fallback location (I think the support for ignition failure is working)

**Comment by Bastrabun** [comment]
Hm. I managed to portal into an area that did not belong to me. For finding the surface, you could use the same mechanic that is used when minetest_game searches for a suitable spawn position: https://github.com/minetest/minetest_game/blob/master/mods/spawn/init.lua

And you'd need to check all nodes the schematic will replace for areas or non-natural nodes, else someone could damage edges and corners.

Then what is natural depends on the game/mapgen/... used. 

We'd probably need to have settings for all that stuff.

nether.nodes_that_are_considered_natural = {"dirt_with_grass", "group:sand", ...}
nether.fail_if_no_position_found = true

Maybe its just me ( and my server), but most protections are on the surface, where players build houses and such. Placing a portal underneath a largely protected area would not do much good. A player trapped in the Nether might have to take a long walk. Usually player protections on the surface do not reach *too* deep, as they usually aren't allowed to protect huge places.

**Comment by Treer** [comment]
It does use the same mechanic that is used when minetest_game searches for a suitable spawn position, but that mechanic is limited - usually returning "not suitable spawn" if the surface is higher than 16 etc, which it often is. A more elaborate search for the surface could be performed here, as you suggest.

All nodes the schematic will replace are checked for protection or non-natural nodes.

Nodes with `is_ground_content` set true are counted as natural, as well as nodes in the tree and leave groups - which is not the same as wood, everything else is assumed unnatural.

If players get themselves trapped under a protected area, they already have fast travel back in the nether. This might be trying to outwit emergent gameplay instead of letting it blossom and see what happens :)

**Comment by Treer** [comment]
> I managed to portal into an area that did not belong to me

Could you post the log?

**Comment by Bastrabun** [comment]
From what I recall: I went with the admin player to a place with unnatural nodes, created a portal to the Nether, checked to see if it worked, went back up and dropped obsidian and mese, then logged out. Logged in with my Testaccount (Violetta) to gather the materials, went into the portal to the Nether. Down there I built another portal with Violetta, ignited and it took me to -253 deep under the surface, but still within the protection.

[portal_into_protection-4.txt](https://github.com/minetest-mods/nether/files/4993398/portal_into_protection-4.txt)
